ai_server/server.py

```diff
@@ -1,12 +1,250 @@
+    """Check if model is available in llama.cpp."""
+    return resolve_model_path(model) is not None
+
+def chat_with_ollama(model: str, content: str) -> str:
+    """Handle chat using ollama."""
+    response = ollama.chat(
+        model=model,
+        messages=[{'role': 'user', 'content': content}],
+        stream=False
+    )
+    return response.message.content
+
+def chat_with_llamacpp(model: str, content: str, timeout: int = 300) -> str:
+    """Handle chat using llama.cpp CLI."""
+    model_path = resolve_model_path(model)
+    
+    if not model_path:
+        available_models = get_available_llamacpp_models()
+        raise ValueError(f"Model not found: {model}. Available models: {available_models}")
+    
+    cmd = [
+        LLAMA_CPP_CLI,
+        '-m', model_path,
+        '--n-gpu-layers', '40',
+        '-p', content,
+        '-n', '512',
+        '--no-display-prompt',
+        '-no-cnv'
+    ]
+    
+    try:
+        result = subprocess.run(
+            cmd,
+            capture_output=True,
+            text=False,
+            timeout=timeout,
+            check=True
+        )
+        
+        # For V3 IQ1_M and IQ2_XXS models
+        try:
+            stdout_text = result.stdout.decode('utf-8')
+        except UnicodeDecodeError:
+            stdout_text = result.stdout.decode('utf-8', errors='replace')
+
+        output_lines = stdout_text.strip().split('\n')
+        response_lines = []
+        for line in output_lines:
+            if any(skip_pattern in line for skip_pattern in LLAMA_CPP_SKIP_PATTERNS):
+                continue
+            
+            if line.startswith('llama_perf_'):
+                break
+                
+            if line.strip():
+                response_lines.append(line.strip())
+        
+        response = ' '.join(response_lines) if response_lines else ""
+        return response if response else "No response generated."
+        
+    except subprocess.TimeoutExpired:
+        raise Exception(f"Llama.cpp request timed out for model {model}")
+    except subprocess.CalledProcessError as e:
+        stderr_text = ""
+        if e.stderr:
+            try:
+                stderr_text = e.stderr.decode('utf-8')
+            except UnicodeDecodeError:
+                stderr_text = e.stderr.decode('utf-8', errors='replace')
+        raise Exception(f"Llama.cpp failed for {model}: {stderr_text.strip() if stderr_text else 'Unknown error'}")
+    except FileNotFoundError:
+        raise Exception("Llama.cpp CLI not found")
+
+def chat_with_model(model: str, content: str, llama_mode: str = "cli") -> str:
+    """Route chat request based on llama_mode: server (external), cli, or default fallback."""
+    if llama_mode == "server":
+        # Use external llama-server with fallbacks
+        if not LLAMA_SERVER_URL:
+            raise Exception("LLAMA_SERVER_URL environment variable not set for server mode")
+        
+        if is_llamacpp_available(model):
```

Comment:
Overall, you're trying to include many fallbacks, which is nice but makes the code more complex. In general, the logic here should be (1) select one thing to try, and (2) if that fails, tell the user. This is especially true when the user has specified an explicit instruction like `llama_mode = 'server'`; it's surprising to then return the response from trying something else.

You can orient your code as:

```
if is_llamacpp_available(model):
    if llama_mode == "server":
        ...
    elif llama_mode == "cli":
        ...
    else:
        error
else:
    # use ollama
```